### PR TITLE
Reduce rustbuild bloat caused by serde_derive

### PR DESCRIPTION
**Pull request by bjorn3**
This reduces the size of the `.text` section from 10.1MiB (6.2MiB for just rustbuild code) to 9.3MiB (5.3MiB for just rustbuild code).

This also reduces compile time from ~6.1s for incr recompilation to ~5.6s.

r? @Mark-Simulacrum

### CODE REVIEW

File: src/bootstrap/config.rs [hunk 1]
```diff
@@ -380,115 +380,201 @@ macro_rules! derive_merge {
...
+                        while let Some(key) =
+                            match serde::de::MapAccess::next_key::<FieldName>(&mut map) {
+                                Ok(val) => val,
+                                Err(err) => {
+                                    return Err(err);
+                                }
+                            }
+                        {
+                            match key {
+                                $(FieldName::$field => {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/config.rs:445]: It seems like the mapping to the enum isn't actually necessary? next_key could just deserialize &str and our match here return the unknown_field if necessary?

Might help further reduce the generated code.

**Review by Mark-Simulacrum — COMMENTED** [review]
Generally seems OK, though it does seem unfortunate to lose the nicer ergonomics offered by serde-derive. Ultimately we're not really using them so it's probably OK.

I am a little surprised to see that the dependency on serde derive isn't entirely dropped -- is that because it's still being used by metadata or something else? Do you expect to drop that eventually?

**Comment by bjorn3** [comment]
> I am a little surprised to see that the dependency on serde derive isn't entirely dropped -- is that because it's still being used by metadata or something else? Do you expect to drop that eventually?

It is still used in a couple of other places like cargo artifact message deserialization and toolstate serialization/deserialization. I want to remove these too, but I focused on config deserialization first as it has the best cost vs benefit ratio.

File: src/bootstrap/config.rs (hunk 1 above)
Comment (bjorn3) [reply src/bootstrap/config.rs]: Done

**Comment by Mark-Simulacrum** [comment]
FWIW, if we care about disk space, it's probably worth comparing against an optimized build (at least -O1) of bootstrap and see what the size vs. time tradeoff is.

But regardless, this seems good to me -- r=me if you don't want to iterate further in this PR.

**Comment by bjorn3** [comment]
My main motivation for reducing this bloat is actually reducing compilation time. Smaller binaries generally take less time to codegen and less time to link. Reducing the total disk usage is a nice side-effect.

By the way final stats for text section size: 4.9MiB for just rustbuild code. 8.8MiB for the whole binary.

@bors r=Mark-Simulacrum

**Pull request merged**